### ISSUE TITLE
Problem running the standardizedSolution_boot_ci() function

### ISSUE THREAD

**Issue by kachunwu** [sfcheung/semhelpinghands#74]
When I try to standardize the bootstrap solution from lavaan, I encounter this error message:

> Error in if (all(abs(out_all[, x] - mean(out_all[, x], na.rm = TRUE)) <  : 
  missing value where TRUE/FALSE needed

Is it because of the bootstrap draw that did not converged? 

This is the command I used to fit the model:
`fit <- sem(model, data = dataset, estimator = "ML", se = "bootstrap",bootstrap = 2000, parallel="snow",ncpus = 16, iseed=1234, fixed.x = FALSE)`

**Comment by kachunwu** [comment]
Thank you for the quick response. Sorry that I cannot share the data and the model but I did check the boot oject using lavInspect(fit, "boot"). There is no single column (paths) with NA for all bootstrap sample, but only few rows (the nonconverge sample).

lavaan version: 0.6-17
semTools version: 0.5-6
semhelpinghands version: 0.1.10

Not sure if it is relevant, but when fitting my model I received the following:
```
> Warning messages:
> 1: In lav_model_nvcov_bootstrap(lavmodel = lavmodel, lavsamplestats = lavsamplestats,  :
>   lavaan WARNING: 17 bootstrap runs failed or did not converge.
> 2: In lav_model_nvcov_bootstrap(lavmodel = lavmodel, lavsamplestats = lavsamplestats,  :
>   lavaan WARNING: 4983 bootstrap runs resulted in nonadmissible solutions.
> 3: In lav_object_post_check(object) :
>   lavaan WARNING: the covariance matrix of the residuals of the observed
>                 variables (theta) is not positive definite;
>                 use lavInspect(fit, "theta") to investigate. 
```

I have tried another functions like the following and successfully plotted the histogram and QQ plot:
```
fit_with_boot_std <- store_boot_est_std(fit)
plot_boot(fit.full, "p1", standardized = TRUE)
```

But when I run standardizedSolution_boot_ci() it still gives the same error message.
```
standardizedSolution_boot_ci(fit_with_boot_std)
> Error in if (all(abs(out_all[, x] - mean(out_all[, x], na.rm = TRUE)) <  : 
>   missing value where TRUE/FALSE needed
```

Since the error happen in the following function of **standardizedSolution_boot_ci.R**, I am thinking is there something missing in my fitted lavaan object that is required by this function.
```
# Adapted from boot
    boot_ci <- sapply(seq_along(est_org), function(x) {
                          if (all(abs(out_all[, x] -
                                      mean(out_all[, x], na.rm = TRUE)) <
                                      1e-8) ||
                              all(is.na(out_all[, x]))) {
                              return(c(NA, NA))
                            }
                          boot::boot.ci(boot_tmp,
                                index = x,
                                type = "perc",
                                conf = level)$percent[4:5]
                        })
```

I am seeking alterantive ways to do it like lavaan::bootstrapLavaan(). Hope to update you about my progress.

**Comment by sfcheung** [comment]
Thanks for the information, which is useful. I will see how to reprodue the error on my end.

By the way, you are right. You can use `lavaan::bootstrapLavaan()` too. This is an illustration:

https://groups.google.com/g/lavaan/c/wkI3f_SPubc/m/_JjYZNwcDwAJ

You may want to add `se = FALSE` in the call to `standardizedSolution()`, to prevent computing the SEs, which you won't need in bootstrapping. It can be slow to run if you did not add `se = FALSE`.

I wrote `standardizedSolution_boot_ci()` to reuse the bootstrap estimates already generated when running `sem()` or `cfa()` with `se = "bootstrap"`. Using `lavaan::bootstrapLavaan()` will refit the models many many times, even if bootstrapping has already been done once. This is inefficient and can take a long time when bootstrapping and FIML are used together. However, if time is not a concern, then `lavaan::bootstrapLavaan()` is enough for the task.

**Comment by sfcheung** [comment]
A quick question, not directly related to `standardizedSolution_boot_ci()` (though may be related indirectly as some bootstrap replications may not have error but the solution is still nonadmissible):

You mention the following warning:

```
> 3: In lav_object_post_check(object) :
>   lavaan WARNING: the covariance matrix of the residuals of the observed
>                 variables (theta) is not positive definite;
>                 use lavInspect(fit, "theta") to investigate. 
```

Will you have this warning if you do not use `se = "bootstrap"`? If yes, then something's wrong when fitting the model. This warning need to be addressed first.

**Comment by sfcheung** [comment]
Regarding the other two functions you mentioned:

```
fit_with_boot_std <- store_boot_est_std(fit)
plot_boot(fit.full, "p1", standardized = TRUE)
```

`store_boot_est_std()` stores the bootstrap estimates for all parameters, including user-defined parameters created by `:=`.

`store_boot_def()`, on the other hand, stores only the bootstrap estimates of the user-defined parameters in the unstandardized solution. Therefore, setting `standardized = TRUE` will result in an error if only `store_boot_def()` was run:

```
fit_with_boot_def <- store_boot_def(fit.full)
plot_boot(fit_with_boot_def, "total1", standardized = TRUE)
> Error in plot_boot(fit_with_boot_def, "total1", standardized = TRUE) : 
>   Bootstrap estimates not found or not stored.
```

Therefore, if you only need to plot the bootstrap estimates of the standardized solution, then using `store_boot_est_std()` is enough, even for the user-defined parameters:

``` r
library(lavaan)
#> This is lavaan 0.6-17
#> lavaan is FREE software! Please report any bugs.
library(semhelpinghands)

set.seed(1234)
mod0 <-
"
x1 ~~ .3*x2
x1 ~~ .15*x3
x2 ~~ .4*x3
"
dat <- simulateData(model = mod0,
                    sample.nobs = 50,
                    empirical = TRUE)
mod <-
"
x2 ~ a*x1
x3 ~ b*x2
ab := a*b
"
set.seed(1234)
suppressWarnings(system.time(fit <- sem(mod,
                       data = dat,
                       se = "boot",
                       bootstrap = 100)))
#>    user  system elapsed 
#>    0.42    0.03    0.77

fit_with_boot_std <- store_boot_est_std(fit)
plot_boot(fit_with_boot_std, "a", standardized = TRUE)
```

![](https://i.imgur.com/avLQSWv.png)

``` r
plot_boot(fit_with_boot_std, "ab", standardized = TRUE)
```

![](https://i.imgur.com/Uodmhqo.png)

<sup>Created on 2024-03-27 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

I think you used those functions just to help figuring out the source of the error (thanks for that) but your example on these functions also shows that I need to come up with a better error message. That one can be misleading. Thanks.

**Comment by sfcheung** [comment]
I guess this is one possible cause of the original error: There are bootstrap samples in which the unstandardized solution has no problem but some parameters cannot be computed or have invalid values (`Inf` or `NaN`) in the standardized solution. This type of cases is rare but possible, and cannot be discovered by insepcting the results of `lavInspect()` with `"boot"`.

I did a quick fix (Version 0.1.10.1) to handle this situation. You can install the latest version from GitHub:

```r
remotes::install_github("sfcheung/semhelpinghands")
```

Please use this vesion and see whether you still encounter the error. Thanks in advance.

I need more time to think about how to handle this case in the `plot_boot()` function and will update it soon.

**Comment by sfcheung** [comment]
The latest version, 0.1.11, with this issue addressed is now on CRAN.
